source/common/http/async_client_impl.cc

```diff
@@ -87,6 +90,18 @@ void AsyncStreamImpl::encodeHeaders(HeaderMapPtr&& headers, bool end_stream) {
 void AsyncStreamImpl::encodeData(Buffer::Instance& data, bool end_stream) {
   ENVOY_LOG(trace, "async http request response data (length={} end_stream={})", data.length(),
             end_stream);
+  // It's possible we have been reset by the client but the router is still
+  // sending data on its way out. This happens for example when
+  // Utility::sendLocalReply() tells the gRPC async client that it has a non-200
+  // via encodeHeaders(), gRPC async client resets the HTTP async stream but on
+  // return to Utility::sendLocalReply() it unconditionally sends a body with
+  // encodeData().
+  // TODO(htuch): This is the only case we catch today, but conceivably we could
+  // see encodeTrailers() as well (or encodeHeaders()?) from router on its way
```

Comment:
Yeah, I think we should ship the fix and test, will add a `TODO` and look into fixing the broader problem today, it seems dodgy to leave it for much longer. 

For coverage, to do unit tests I need to make the stream decoder callbacks on `Http::AsycnStreamImpl` public to poke at via the unit test. I notice we do this in various places, seems not so great though to have to change visibility of entire interfaces just for unit tests.